### DISCUSSION TITLE
Extension, Explorer Tree Items, and Data Attributes

### DISCUSSION THREAD

**Discussion by shellscape** [microsoft/vscode-discussions#218] [Extension Development QnA] [answered]
Hey all. I'm on a (what some would call futile and silly) quest to completely rework the VS Code UX to match that of Atom. It's my personal view that Atom's UX is more polished, so I'm going the effective skinning route to make that happen.

During my work on this, I've noticed that the various `DecorationProvider` classes in use by extensions like Git and TypeScript are able to add decoration classes which define certain properties like `color`. However, this is problematic for anyone inspecting the class names. It wouldn't be so bad if not for a few things, which ultimately I think were poor design choices:

1. The tree structure is flat, rather than nested
2. Each "level" of the tree receives a different decoration class name. e.g. `.monaco-decoration-iconBadge--9ieymw`.
3. Nothing but title is modified to indicate a "state" of the file wrt the git repo. e.g `title="~/code/thing.ts • Untracked"`

It seems to me that there could be a number of uses for a data attribute on the element such like `data-ext-git-state="untracked"`. This would allow for a wider breadth of styling, like changing the tree icon based on git state, or throwing a squiggly line under files which contain errors. Right now we're relegated to CSS like this:

```
.monaco-icon-label[title~='Modified']::before {
  filter: brightness(0) saturate(100%) invert(84%) sepia(14%) saturate(912%) hue-rotate(351deg) brightness(94%) contrast(88%) !important;
}
```

That css changes the icon color to match the foreground color for modified files. It's not ideal, and it's admittedly brittle as the title attribute may change in any subsequent release of the extension. Since I don't think it would be reasonable to suggest that the decoration class code needs to be changed, I do think there's benefit in either allowing extensions to add class names, or data attribute names (preferably).

Is there anything in place that would allow for this? If not, what does the path for adding this capability look like? I've never contributed to VS Code before, and while I'm a seasoned engineer, please do reply as if I know little about the structure of VS Code internals and extension authoring. Cheers

**Comment by burkeholland** [comment, marked as answer]
Circling back on this - did you ever find an answer to your question?

**Reply by shellscape** [reply]
Nope. And the dev team summarily shut down a related issue without any decent discussion. See: https://github.com/microsoft/vscode/issues/165753

I've moved onto VS Codium as the project is more aligned with users and has reverted several malicious changes to VS Code. My interaction with the team maintaining VS Code has left me of the opinion that Microsoft are terrible stewards of open source.

**Comment by isidorn** [comment]
Closing as this is duplicate of https://github.com/microsoft/vscode/issues/165753

**Reply by shellscape** [reply]
lol this team is hilarious
